### DISCUSSION TITLE
from_self() not populating relationships correctly when querying polymorphic entity relations

### DISCUSSION THREAD

**Discussion by torsdag** [sqlalchemy/sqlalchemy#7427] [Usage Questions]
### Describe the bug

I am running into an issue where using from_self() results in the relationships of the entity coming out incorrect. It appears to be related to multiple polymorphic relationships being loaded. I have not yet fully understood what is going on but below is a reproducible example. 

Am I doing something stupid or is this a bug?

### To Reproduce

```python
import sqlalchemy.orm

from sqlalchemy.ext.declarative import (
    declarative_base
)

Base = declarative_base()


class Entity(Base):
    __tablename__ = 'entity'

    id = sqlalchemy.Column(
        sqlalchemy.Integer, primary_key=True
    )

    parent_id = sqlalchemy.Column(
        sqlalchemy.Integer, sqlalchemy.ForeignKey(id), nullable=True
    )

    children = sqlalchemy.orm.relationship(
        'Entity', backref=sqlalchemy.orm.backref(
            'parent',  remote_side='Entity.id'
        )
    )

    attribute_type = sqlalchemy.Column(
        sqlalchemy.String(32)
    )

    __mapper_args__ = {
        'polymorphic_on': attribute_type,
        'polymorphic_identity': 'entity'
    }


class EntityTwo(Entity):
    __tablename__ = 'entity_two'

    id = sqlalchemy.Column(
        sqlalchemy.Integer, sqlalchemy.ForeignKey(Entity.id), primary_key=True
    )

    name = sqlalchemy.Column(
        sqlalchemy.String(32)
    )

    __mapper_args__ = {
        'polymorphic_identity': 'entity_two'
    }


engine = sqlalchemy.create_engine(
    'sqlite:///:memory:', echo=True
)

Base.metadata.create_all(engine)

Session = sqlalchemy.orm.sessionmaker(
    bind=engine
)

session = Session()

session.add(
    EntityTwo(
        id=1,
        name='test1'
    )
)
session.add(
    EntityTwo(
        id=2,
        parent_id=1
    )
)

session.commit()


result = session.query(
    EntityTwo
).options(
    sqlalchemy.orm.joinedload(
        Entity.parent
    )
).options(
    sqlalchemy.orm.joinedload(
        Entity.children
    )
).filter(
    Entity.id == 2
)

# works as expected
assert(result.one().parent is not None)

# fails when using from_self()
assert(result.from_self().one().parent is not None)
```

### Error

Traceback (most recent call last):
  File "test.py", line 99, in <module>
    assert(result.from_self().one().parent is not None)
AssertionError

### Versions

- OS: fedora 34
- Python: 2.7.18
- SQLAlchemy: 1.4.27
- Database: sqlite (tested), mariadb ( tested )

**Comment by zzzeek** [comment]
hi -

from_self() is deprecated and will not be in the next major version of SQLAlchemy.  can you approach your problem using a different technique?    background at https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#selecting-from-the-query-itself-as-a-subquery-e-g-from-self  .

**Comment by torsdag** [comment]
Hey, thank you for the quick reply!

I have attempted using `sqlalchemy.orm.aliased(EntityTwo, result.subquery())` but opposed to `from_self()` the selected columns/joins are not kept. Is there a way of replicating that behavior in `from_self()` using aliased?

```python
result = session.query(
    EntityTwo
).options(
    sqlalchemy.orm.joinedload(
        Entity.parent
    )
).options(
    sqlalchemy.orm.joinedload(
        Entity.children
    )
).filter(
    Entity.id == 2
)

aliased_result = session.query(
    sqlalchemy.orm.aliased(
        EntityTwo, result.subquery()
    )
)

print(str(result.from_self()))
#SELECT anon_1.entity_two_id AS anon_1_entity_two_id, anon_1.entity_parent_id AS anon_1_entity_parent_id, anon_1.entity_attribute_type AS anon_1_entity_attribute_type, anon_1.entity_two_name AS anon_1_entity_two_name, entity_1.id AS entity_1_id, entity_1.parent_id AS entity_1_parent_id, entity_1.attribute_type AS entity_1_attribute_type, entity_2.id AS entity_2_id, entity_2.parent_id AS entity_2_parent_id, entity_2.attribute_type AS entity_2_attribute_type 
FROM (SELECT entity_two.id AS entity_two_id, entity.id AS entity_id, entity.parent_id AS entity_parent_id, entity.attribute_type AS entity_attribute_type, entity_two.name AS entity_two_name 
FROM entity JOIN entity_two ON entity.id = entity_two.id 
WHERE entity.id = ?) AS anon_1 LEFT OUTER JOIN entity AS entity_1 ON anon_1.entity_two_id = entity_1.parent_id LEFT OUTER JOIN entity AS entity_2 ON entity_2.id = entity_1.parent_id


print(str(aliased_result))
#SELECT anon_1.id AS anon_1_id, anon_1.parent_id AS anon_1_parent_id, anon_1.attribute_type AS anon_1_attribute_type, anon_1.name AS anon_1_name 
FROM (SELECT entity_two.id AS id, entity.id AS id_1, entity.parent_id AS parent_id, entity.attribute_type AS attribute_type, entity_two.name AS name 
FROM entity JOIN entity_two ON entity.id = entity_two.id 
WHERE entity.id = ?) AS anon_1

```

cheers

**Comment by zzzeek** [comment]
so session.query(anything_aliased(EntityTwo)) is going to look only for columns it needs to load an EntityTwo object.  if you are looking for it to also ask for columns that represent its related elements loaded for joinedload() you would use [contains_eager](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html?highlight=contains_eager#sqlalchemy.orm.contains_eager) to set those up, if that's what you're going for.

```py
e2a = sqlalchemy.orm.aliased(
        EntityTwo, result.subquery()
    )

aliased_result = session.query(e2a).options(contains_eager(e2a.parent), contains_eager(e2a.children))
```

these are very advanced queries where things frequently don't work completely so if that isn't working as I think it should I can try to dig in further.

**Comment by torsdag** [comment]
Hey again, 

I am having no success with `contains_eager` as it sadly appears to do nothing in this case.

If it were to work is it feasible to gather the required columns / joins from the original query and apply them again using this technique? 

```python
e2a = sqlalchemy.orm.aliased(
    EntityTwo, result.subquery()
)

aliased_result1 = session.query(
    e2a
)

aliased_result2 = aliased_result1.options(
    sqlalchemy.orm.contains_eager(
        e2a.parent
    ),
    sqlalchemy.orm.contains_eager(
        e2a.children
    )
)

print(str(aliased_result1))
SELECT anon_1.id AS anon_1_id, anon_1.parent_id AS anon_1_parent_id, anon_1.attribute_type AS anon_1_attribute_type, anon_1.name AS anon_1_name 
FROM (SELECT entity_two.id AS id, entity.id AS id_1, entity.parent_id AS parent_id, entity.attribute_type AS attribute_type, entity_two.name AS name 
FROM entity JOIN entity_two ON entity.id = entity_two.id 
WHERE entity.id = ?) AS anon_1

print(str(aliased_result2))
SELECT anon_1.id AS anon_1_id, anon_1.parent_id AS anon_1_parent_id, anon_1.attribute_type AS anon_1_attribute_type, anon_1.name AS anon_1_name 
FROM (SELECT entity_two.id AS id, entity.id AS id_1, entity.parent_id AS parent_id, entity.attribute_type AS attribute_type, entity_two.name AS name 
FROM entity JOIN entity_two ON entity.id = entity_two.id 
WHERE entity.id = ?) AS anon_1
```

cheers

**Reply by zzzeek** [reply]
oh, one other thing is that those joinedload() options aren't going to translate when you make a subquery, which is likely what's going on here.   

beyond that, I have a script here that shows how this should work, but it also does not work due to the deep self-referential things you have, so I cannot fix this right now and you will need to solve your problem differently until i have time to look at this.

new issue is added at #7426, which contains a script that is what should work here, but unfortauntely doesnt as the ORM does not correctly target the "id_1" column in the statement.

**Reply by torsdag** [reply]
thank you for taking a look! will work around it for now

cheers
Eric

**Reply by zzzeek** [reply]
the "contains_eager against a subquery" thing has been coming up in recent weeks so I do want to dig into this and get the cases to be cleaner.
